File: ios/PersonalSite/Features/AiNews/AiNewsView.swift
import SwiftUI

/// AI 动态列表：直连 ai_news_public_items，按北京时间日分组，倒序分页。
@MainActor
@Observable
final class AiNewsListModel {
    private(set) var groups: [AiNewsDayGroup<AiNewsListItem>] = []
    private(set) var isLoading = false
    private(set) var errorMessage: String?
    private var offset = 0
    private var hasMore = true

    private static let pageSize = 50
    // 与 Web 端相同的 PostgREST 投影：content jsonb 字段别名平铺成行。
    private static let listSelect = "category:content->>category,id,publishedAt:content->>publishedAt,selected,sourceName:content->>sourceName,summary:content->>summary,title:content->>title"

    var isEmpty: Bool { groups.isEmpty }

    func loadInitial() async {
        guard groups.isEmpty, !isLoading else { return }
        await load(reset: true)
    }

    func refresh() async {
        await load(reset: true)
    }

    func loadMoreIfNeeded(currentItem: AiNewsListItem) async {
        let allItems = groups.flatMap(\.items)
        guard hasMore, !isLoading, currentItem.id == allItems.last?.id else { return }
        await load(reset: false)
    }

    private func load(reset: Bool) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        if reset {
            offset = 0
            hasMore = true
        }
        do {
            // 多取一条判断 hasMore，与 Web 的 getAiNewsPage 一致。
            let rows: [AiNewsListItem] = try await SupabaseClientProvider.shared
                .from("ai_news_public_items")
                .select(Self.listSelect)
                .order("published_at", ascending: false, nullsFirst: false)
                .range(from: offset, to: offset + Self.pageSize)
                .execute()
                .value
            hasMore = rows.count > Self.pageSize
            let pageItems = rows.prefix(Self.pageSize)
            let merged = reset ? Array(pageItems) : groups.flatMap(\.items) + pageItems
            offset += pageItems.count
            groups = AiNewsGrouping.group(merged) { $0.publishedAt }
        } catch {
            errorMessage = "读取每日动态失败，请检查网络后重试。"
        }
    }
}

struct AiNewsView: View {
    @State private var model = AiNewsListModel()
    var headerCollapsed: Binding<Bool>?

    private let scrollSpace = "ai-news-list-scroll"

    init(headerCollapsed: Binding<Bool>? = nil) {
        self.headerCollapsed = headerCollapsed
    }

    var body: some View {
        NavigationStack {
            LoadStateView(
                isLoading: model.isLoading && model.isEmpty,
                errorMessage: model.isEmpty ? model.errorMessage : nil,
                isEmpty: model.isEmpty,
                emptyMessage: "暂无每日动态",
                onRetry: { Task { await model.refresh() } }
            ) {
                List {
                    if let headerCollapsed {
                        ScrollCollapseSensor(
                            coordinateSpace: scrollSpace,
                            isCollapsed: headerCollapsed
                        )
                        .listRowInsets(.init())
                        .listRowSeparator(.hidden)
                        .listRowBackground(Color.clear)
                    }
                    AiNewsStreamContent(model: model)
                }
                .listStyle(.plain)
                .environment(\.defaultMinListRowHeight, 1)
                .scrollContentBackground(.hidden)
                .coordinateSpace(.named(scrollSpace))
                .trackHeaderCollapse(headerCollapsed)
                .refreshable { await model.refresh() }
            }
            .toolbar(.hidden, for: .navigationBar)
            .background(Color.psSurface)
            .navigationDestination(for: String.self) { id in
                AiNewsDetailView(id: id)
            }
        }
        .task { await model.loadInitial() }
    }
}

/// AI 动态内容流：日分组 + 行 + 滚动到底分页，供 AiNewsView（List）与
/// 首页（ScrollView 内嵌）共用；下钻由外层 NavigationStack 的
/// navigationDestination(for: String.self) 提供。
struct AiNewsStreamContent: View {
    let model: AiNewsListModel

    var body: some View {
        ForEach(model.groups, id: \.dayKey) { group in
            Section {
                ForEach(group.items) { item in
                    NavigationLink(value: item.id) {
                        AiNewsRow(item: item)
                    }
                    .contentListRowChrome()
                    .onAppear {
                        Task { await model.loadMoreIfNeeded(currentItem: item) }
                    }
                }
            } header: {
                HStack(alignment: .firstTextBaseline, spacing: 6) {
                    Text(group.label)
                        .font(.system(size: 14.5, weight: .semibold))
                        .foregroundStyle(Color.psInk)
                    if !group.weekday.isEmpty {
                        Text(group.weekday)
                            .font(.system(size: ContentListMetrics.metadataSize, weight: .medium))
                            .foregroundStyle(Color.psQuiet)
                    }
                    Spacer(minLength: 0)
                }
                .textCase(nil)
                .padding(.top, 10)
                .padding(.bottom, 2)
                .background(Color.psSurface)
            }
        }
    }
}

struct AiNewsRow: View {
    let item: AiNewsListItem

    var body: some View {
        VStack(alignment: .leading, spacing: ContentListMetrics.rowSpacing) {
            ContentListMetadataLine(items: [
                AiNewsCategory.label(for: item.category),
                AiNewsGrouping.relativeTime(for: item.publishedAt) ?? "",
            ])
            Text(item.title)
                .contentListTitle()
                .lineLimit(2)
            if !item.summary.isEmpty {
                Text(item.summary)
                    .contentListSummary()
            }
        }
        .contentListBody()
    }
}

File: android/app/src/main/java/com/personalsite/features/ainews/AiNewsView.kt
package com.personalsite.features.ainews

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Text
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.personalsite.core.PSColors
import com.personalsite.core.SupabaseClientProvider
import com.personalsite.features.support.ContentListMetadataLine
import com.personalsite.features.support.ContentListMetrics
import com.personalsite.features.support.LoadStateView
import com.personalsite.features.support.contentListBody
import com.personalsite.models.AiNewsCategory
import com.personalsite.models.AiNewsDayGroup
import com.personalsite.models.AiNewsGrouping
import com.personalsite.models.AiNewsListItem
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.launch

/** AI 动态列表：直连 ai_news_public_items，按北京时间日分组，倒序分页。 */
class AiNewsListModel : ViewModel() {
    var groups by mutableStateOf<List<AiNewsDayGroup<AiNewsListItem>>>(emptyList())
        private set
    var isLoading by mutableStateOf(false)
        private set
    var isRefreshing by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    private var offset = 0
    private var hasMore = true

    val isEmpty: Boolean get() = groups.isEmpty()

    fun loadInitial() {
        if (groups.isNotEmpty() || isLoading) return
        load(reset = true)
    }

    fun refresh() {
        if (isRefreshing) return
        isRefreshing = true
        load(reset = true) { isRefreshing = false }
    }

    fun loadMoreIfNeeded(currentItemId: String) {
        val allItems = groups.flatMap { it.items }
        if (!hasMore || isLoading || currentItemId != allItems.lastOrNull()?.id) return
        load(reset = false)
    }

    private fun load(reset: Boolean, onComplete: (() -> Unit)? = null) {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null
            try {
                if (reset) {
                    offset = 0
                    hasMore = true
                }
                // 多取一条判断 hasMore，与 Web 的 getAiNewsPage 一致。
                // 与 Web 端相同的 PostgREST 投影：content jsonb 字段别名平铺成行。
                val rows = SupabaseClientProvider.shared
                    .from("ai_news_public_items")
                    .select(columns = Columns.raw(LIST_SELECT)) {
                        order("published_at", Order.DESCENDING, nullsFirst = false)
                        range(offset.toLong(), (offset + PAGE_SIZE).toLong())
                    }
                    .decodeList<AiNewsListItem>()
                hasMore = rows.size > PAGE_SIZE
                val pageItems = rows.take(PAGE_SIZE)
                val merged = if (reset) pageItems else groups.flatMap { it.items } + pageItems
                offset += pageItems.size
                groups = AiNewsGrouping.group(merged) { it.publishedAt }
            } catch (_: Exception) {
                errorMessage = "读取每日动态失败，请检查网络后重试。"
            } finally {
                isLoading = false
                onComplete?.invoke()
            }
        }
    }

    companion object {
        private const val PAGE_SIZE = 50
        private const val LIST_SELECT =
            "category:content->>category,id,publishedAt:content->>publishedAt,selected,sourceName:content->>sourceName,summary:content->>summary,title:content->>title"
    }
}

/** 每日动态列表页：日分组 Section + 行 + 滚动到底分页 + 下拉刷新（对齐 iOS AiNewsView）。 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AiNewsView(
    onOpenDetail: (String) -> Unit,
    onHeaderCollapse: (Boolean) -> Unit,
    model: AiNewsListModel = viewModel(),
) {
    val listState = rememberLazyListState()

    // 滚动超过 ~20dp 时折叠页头，回落时恢复（对齐 iOS ScrollCollapseSensor）
    val collapseThreshold = with(LocalDensity.current) { 20.dp.roundToPx() }
    val headerCollapsed by remember {
        derivedStateOf {
            listState.firstVisibleItemIndex > 0 ||
                listState.firstVisibleItemScrollOffset > collapseThreshold
        }
    }
    LaunchedEffect(headerCollapsed) { onHeaderCollapse(headerCollapsed) }

    // 最后一条数据行可见时触发分页（对齐 iOS 行 onAppear 里的 loadMoreIfNeeded）
    val lastVisibleItemId by remember {
        derivedStateOf {
            val info = listState.layoutInfo
            val lastIndex = info.totalItemsCount - 1
            if (lastIndex >= 0 && info.visibleItemsInfo.any { it.index == lastIndex }) {
                model.groups.flatMap { it.items }.lastOrNull()?.id
            } else {
                null
            }
        }
    }
    LaunchedEffect(lastVisibleItemId) {
        lastVisibleItemId?.let { model.loadMoreIfNeeded(it) }
    }

    LaunchedEffect(Unit) { model.loadInitial() }

    LoadStateView(
        isLoading = model.isLoading && model.isEmpty,
        errorMessage = if (model.isEmpty) model.errorMessage else null,
        isEmpty = model.isEmpty,
        emptyMessage = "暂无每日动态",
        onRetry = { model.refresh() },
    ) {
        PullToRefreshBox(
            isRefreshing = model.isRefreshing,
            onRefresh = { model.refresh() },
            modifier = Modifier.fillMaxSize(),
        ) {
            LazyColumn(
                state = listState,
                modifier = Modifier
                    .fillMaxSize()
                    .background(PSColors.surface),
            ) {
                model.groups.forEach { group ->
                    item(key = "header:${group.dayKey}") {
                        AiNewsDayHeader(group = group)
                    }
                    items(group.items, key = { it.id }) { item ->
                        AiNewsRow(item = item, onClick = { onOpenDetail(item.id) })
                    }
                }
            }
        }
    }
}

/** 日分组头：日期 label + 中文 weekday（对齐 iOS Section header）。 */
@Composable
private fun AiNewsDayHeader(group: AiNewsDayGroup<AiNewsListItem>) {
    Row(
        verticalAlignment = Alignment.Bottom,
        modifier = Modifier
            .fillMaxWidth()
            .background(PSColors.surface)
            .padding(horizontal = 16.dp)
            .padding(top = 10.dp, bottom = 2.dp),
    ) {
        Text(
            group.label,
            fontSize = 14.5.sp,
            fontWeight = FontWeight.SemiBold,
            color = PSColors.ink,
        )
        if (group.weekday.isNotEmpty()) {
            Spacer(Modifier.width(6.dp))
            Text(
                group.weekday,
                fontSize = ContentListMetrics.metadataSize,
                fontWeight = FontWeight.Medium,
                color = PSColors.quiet,
            )
        }
    }
}

/** 列表行：元信息行（分类 · 相对时间）+ 标题（2 行）+ 摘要（2 行）。 */
@Composable
private fun AiNewsRow(item: AiNewsListItem, onClick: () -> Unit) {
    Column(
        verticalArrangement = Arrangement.spacedBy(ContentListMetrics.rowSpacing),
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(ContentListMetrics.rowPadding)
            .contentListBody(),
    ) {
        ContentListMetadataLine(
            items = listOf(
                AiNewsCategory.label(item.category),
                AiNewsGrouping.relativeTime(item.publishedAt) ?: "",
            ),
        )
        Text(
            item.title,
            fontSize = ContentListMetrics.titleSize,
            fontWeight = FontWeight.SemiBold,
            color = PSColors.ink,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
        if (item.summary.isNotEmpty()) {
            Text(
                item.summary,
                fontSize = ContentListMetrics.summarySize,
                color = PSColors.quiet,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}
